### PR TITLE
Show the last 50 lines of stderr if a crash occurs in instrumentation mode.

### LINKED ISSUE #361: Instrumentation mode doesn't show stderr in the event of a real crash.
CC @fsfod 

When a benchmark crashes, often useful crash information is output to stderr. Under normal circumstances, Krun will capture stderr and print it out in the event of a crash.

Under instrumentation mode, stderr is siphoned off to a file for later parsing and the error reporting is unaware of this, so in the case of a crash, it prints an empty stderr, which isn't helpful.

### Dumb fix

Get Krun to look into the file to print the error.

This is bad in the case of PyPy as it makes a ton of instrumentation output.

### Simple(?) Fix

Most instrumentation parsers use a regex to decided if a line is instrumentation data. In the case of a crash, print only the lines which didn't match.

Other ideas?

### PR DESCRIPTION
**Pull request by vext01**
This is the least instrustive method of fixing #361. It's not perfect, but the other ways touch a lot more code.

What do you think?

Fixes #361

### CODE REVIEW

File: krun/util.py [hunk 1]
```diff
@@ -260,6 +260,22 @@ def check_and_parse_execution_results(stdout, stderr, rc, config,
             err_s += "Exception string: %s\n" % str(e)
         err_s += "return code: %d\n" % rc
         err_s += "stdout:\n%s\n%s\n%s\n\n" % (rule, stdout, rule)
+
+        # In instrumentation mode, stderr will have been written to disk for
+        # parsing. Read the last 50 lines back (it may be huge) so as to offer
+        # the user at-least something relating to the error.
+        if instrument:
+            from krun.vm_defs import INST_STDERR_FILE
```
Comment (ltratt) [inline krun/util.py]: Should this `import` be at the top of the file?

File: krun/util.py [hunk 2]
```diff
@@ -260,6 +260,22 @@ def check_and_parse_execution_results(stdout, stderr, rc, config,
...
+            num_stderr_lines = 0
+            with open(INST_STDERR_FILE) as fh:
+                num_stderr_lines += 1
+            stderr_lines = []
+            with open(INST_STDERR_FILE) as fh:
+                for line_num, line in enumerate(fh):
+                    if line_num >= num_stderr_lines - 50:
```
Comment (ltratt) [inline krun/util.py]: We should make the "50" a constant. I'd also suggest making it more like 100 (it's not totally uncommon to have 50 stack frames).

File: krun/util.py [hunk 3]
```diff
@@ -260,6 +260,22 @@ def check_and_parse_execution_results(stdout, stderr, rc, config,
...
         err_s += "return code: %d\n" % rc
         err_s += "stdout:\n%s\n%s\n%s\n\n" % (rule, stdout, rule)
+
+        # In instrumentation mode, stderr will have been written to disk for
+        # parsing. Read the last 50 lines back (it may be huge) so as to offer
+        # the user at-least something relating to the error.
+        if instrument:
+            from krun.vm_defs import INST_STDERR_FILE
+            num_stderr_lines = 0
+            with open(INST_STDERR_FILE) as fh:
```
Comment (vext01) [inline krun/util.py]: Can I re-use this file handle somehow?

**Comment by vext01** [comment]
Here's the reason for the need for this hack btw:
https://github.com/softdevteam/krun/blob/master/krun/vm_defs.py#L222-L231

File: krun/util.py (hunk 2 above)
Comment (vext01) [reply krun/util.py]: sure

File: krun/util.py (hunk 1 above)
Comment (vext01) [reply krun/util.py]: no, that's the circular import I was moaning about

**Comment by vext01** [comment]
Example crash output:

```
...
[2017-10-17 17:07:41: DEBUG] sync disks...
[2017-10-17 17:07:41: WARNING] SIMULATED: `time.sleep(30)` (--quick)
[2017-10-17 17:07:41: ERROR] Benchmark returned non-zero or emitted invalid JSON.
Exception string: No JSON object could be decoded
return code: 0
stdout:
--------------------------------------------------

--------------------------------------------------

stderr:
--------------------------------------------------
[26a7c55693a494] {gc-set-nursery-size
nursery size: 270336
[26a7c556953c22] gc-set-nursery-size}
[26a7c556961d9e] {gc-hardware
L2cache = 8388608
[26a7c5569a2106] gc-hardware}
[26a7c5569a5ca0] {gc-hardware
memtotal = 25244209152.000000
[26a7c5569b6cdc] gc-hardware}
[26a7c5569b7f32] {gc-minor
[26a7c5569b8d3e] {gc-minor-walkroots
[26a7c5569bbb76] gc-minor-walkroots}
minor collect, total memory used: 0
number of pinned objects: 0
[26a7c5569c033e] gc-minor}
[26a7c5569c5180] {gc-set-nursery-size
nursery size: 4194304
[26a7c5569ca144] gc-set-nursery-size}
[26a7c55785a496] {gc-minor
[26a7c557861faa] {gc-minor-walkroots
[26a7c55787a8fe] gc-minor-walkroots}
minor collect, total memory used: 827800
number of pinned objects: 0
[26a7c557acaf80] gc-minor}
[26a7c558a8fd82] {gc-minor
[26a7c558a9941e] {gc-minor-walkroots
[26a7c558aa5a8a] gc-minor-walkroots}
minor collect, total memory used: 1444808
number of pinned objects: 0
[26a7c558c76730] gc-minor}
[26a7c5593fdbb4] {gc-minor
[26a7c5594025e6] {gc-minor-walkroots
[26a7c559415fc8] gc-minor-walkroots}
minor collect, total memory used: 2258384
number of pinned objects: 0
[26a7c5596b856a] gc-minor}
[26a7c55aa202f4] {gc-minor
[26a7c55aa29198] {gc-minor-walkroots
[26a7c55aa3ece0] gc-minor-walkroots}
minor collect, total memory used: 2780464
number of pinned objects: 0
[26a7c55abfd79e] gc-minor}
[26a7c55c32839e] {gc-minor
[26a7c55c330136] {gc-minor-walkroots
[26a7c55c34277a] gc-minor-walkroots}
minor collect, total memory used: 3103424
number of pinned objects: 0
[26a7c55c450da8] gc-minor}
[26a7c55d8eabc8] {gc-minor
[26a7c55d8eff12] {gc-minor-walkroots
[26a7c55d900720] gc-minor-walkroots}
minor collect, total memory used: 3170512
number of pinned objects: 0
[26a7c55d94bcee] gc-minor}
[26a7c55e8ff80c] {gc-minor
[26a7c55e904d0a] {gc-minor-walkroots
[26a7c55e929fb8] gc-minor-walkroots}
minor collect, total memory used: 3803632
number of pinned objects: 0
[26a7c55eaaf80a] gc-minor}
[26a7c55f911286] {gc-minor
[26a7c55f916b6e] {gc-minor-walkroots
[26a7c5604bf97a] gc-minor-walkroots}
minor collect, total memory used: 4364672
number of pinned objects: 0
[26a7c56060892e] gc-minor}
[26a7c5610a280c] {gc-minor
[26a7c5610a7d14] {gc-minor-walkroots
[26a7c5610bad52] gc-minor-walkroots}
minor collect, total memory used: 5384128
number of pinned objects: 0
[26a7c5612c6f10] gc-minor}
[iterations_runner.py] iteration 1/5
Traceback (most recent call last):
  File "/home/vext01/research/krun/iterations_runners/iterations_runner.py", line 121, in <module>
    bench_func(param)
  File "/home/vext01/research/krun/examples/benchmarks/fannkuch_redux/python/bench.py", line 16, in run_iter
    inner_iter()
  File "/home/vext01/research/krun/examples/benchmarks/fannkuch_redux/python/bench.py", line 65, in inner_iter
    f()
NameError: global name 'f' is not defined

--------------------------------------------------

[2017-10-17 17:07:41: INFO] Finished 'fannkuch_redux(10)' (default-python variant) under 'PyPy'
[2017-10-17 17:07:41: DEBUG] execute shell cmd: /usr/bin/sudo -u krun rm /tmp/envlog-HHk2rw.env
[2017-10-17 17:07:41: DEBUG] Update manifest flag: E
...
```

File: krun/util.py (hunk 1 above)
Comment (ltratt) [reply krun/util.py]: Ah, fine. Maybe add a comment to that effect? Even just `# Avoid circular import problems` or similar.

Comment (ltratt) [reply krun/util.py]: Point taken.

**Comment by vext01** [comment]
Try this.

**Comment by ltratt** [comment]
Works for me! Please squash and I'll merge.

**Comment by vext01** [comment]
Tried to test on Tom's code but I think he has since fixed the LuaJIT error.

**Pull request merged**